### DISCUSSION TITLE
How to get the performance of re2 in regex

### DISCUSSION THREAD

**Discussion by freekeeper** [rust-lang/regex#903] [Q&A] [answered]
I want to compare the performance of regex with re2, 
how to get the result in https://github.com/rust-lang/regex/blob/master/bench/log/05/re2-vs-rust

**Comment by BurntSushi** [comment, marked as answer]
```
$ cd bench
$ ./run rust sherlock::name_alt | tee rust.log
$ ./run re2 sherlock::name_alt | tee re2.log
$ cargo benchcmp re2.log rust.log                                                              
 name                        re2.log ns/iter       rust.log ns/iter     diff ns/iter   diff %  speedup 
 sherlock::name_alt1         46,296 (12850 MB/s)   35,976 (16536 MB/s)       -10,320  -22.29%   x 1.29 
 sherlock::name_alt2         970,626 (612 MB/s)    85,602 (6949 MB/s)       -885,024  -91.18%  x 11.34 
 sherlock::name_alt3         996,394 (597 MB/s)    96,751 (6149 MB/s)       -899,643  -90.29%  x 10.30 
 sherlock::name_alt3_nocase  1,001,320 (594 MB/s)  377,102 (1577 MB/s)      -624,218  -62.34%   x 2.66 
 sherlock::name_alt4         972,360 (611 MB/s)    121,969 (4877 MB/s)      -850,391  -87.46%   x 7.97 
 sherlock::name_alt4_nocase  988,821 (601 MB/s)    184,097 (3231 MB/s)      -804,724  -81.38%   x 5.37 
 sherlock::name_alt5         982,030 (605 MB/s)    90,062 (6605 MB/s)       -891,968  -90.83%  x 10.90 
 sherlock::name_alt5_nocase  981,768 (605 MB/s)    349,229 (1703 MB/s)      -632,539  -64.43%   x 2.81
```

If you remove the `sherlock::name_alt` bit, then all benchmarks will be run.

You can get `cargo benchcmp` by running `cargo install cargo-benchcmp`. Its repo is: <https://github.com/BurntSushi/cargo-benchcmp>.

-----

Putting that aside, I would strongly encourage you to ask more questions. Particularly if you're trying to understand a result. I would also strongly encourage you to benchmark your actual use case, instead of relying on only the benchmarks I've put together. I can also help try to explain differences in perf since I'm pretty familiar with RE2.

**Reply by freekeeper** [reply]
when I run "./run re2", I get the following error, how can I fix it
```
  = note: /usr/bin/ld: /home/freekeeper/Desktop/regex/target/release/build/regex-benchmark-9f348cd2f19de96d/out/libcre2.a(re2.o): in function `re2_regexp_new':
          /home/freekeeper/Desktop/regex/bench/src/ffi/re2.cpp:18: undefined reference to `re2::RE2::RE2(re2::StringPiece const&)'
          /usr/bin/ld: /home/freekeeper/Desktop/regex/target/release/build/regex-benchmark-9f348cd2f19de96d/out/libcre2.a(re2.o): in function `re2_regexp_free':
          /home/freekeeper/Desktop/regex/bench/src/ffi/re2.cpp:22: undefined reference to `re2::RE2::~RE2()'
          /usr/bin/ld: /home/freekeeper/Desktop/regex/target/release/build/regex-benchmark-9f348cd2f19de96d/out/libcre2.a(re2.o): in function `re2_regexp_match':
          /home/freekeeper/Desktop/regex/bench/src/ffi/re2.cpp:30: undefined reference to `re2::RE2::Match(re2::StringPiece const&, unsigned long, unsigned long, re2::RE2::Anchor, re2::StringPiece*, int) const'
          /usr/bin/ld: /home/freekeeper/Desktop/regex/target/release/build/regex-benchmark-9f348cd2f19de96d/out/libcre2.a(re2.o): in function `re2_regexp_find':
          /home/freekeeper/Desktop/regex/bench/src/ffi/re2.cpp:42: undefined reference to `re2::RE2::Match(re2::StringPiece const&, unsigned long, unsigned long, re2::RE2::Anchor, re2::StringPiece*, int) const'
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

warning: `regex-benchmark` (bench "bench") generated 5 warnings
error: could not compile `regex-benchmark` due to previous error; 5 warnings emitted
```

**Reply by freekeeper** [reply]
I tried benchmark the following regular expressions in `regex`
```
bench_captures!(
    rure_dot_star_capture_1MB,
    Regex::new(r"(?s)((.*)()()($))").unwrap(),
    5,
    get_text(TXT_1K, empty_suffix())
);
bench_captures!(
    rure_dot_star_capture_1MB,
    Regex::new(r"(?s)((.*)()()($))").unwrap(),
    5,
    get_text(TXT_1MB, empty_suffix())
);
```
I got the following result
```
test misc::rure_dot_star_capture_1KB         ... bench:      14,615 ns/iter (+/- 679) = 70 MB/s
test misc::rure_dot_star_capture_1MB         ... bench: 189,294,620 ns/iter (+/- 74,675,832) = 5 MB/s
```
when I execute the same test text and regular expressions in `RE2`, I get the following result
```
FullMatch_DotStarCapture_CachedRE2/1K   1000000              2392 ns/op         427.96 MB/s
FullMatch_DotStarCapture_CachedRE2/1M   1000              2020364 ns/op         519.00 MB/s
```
It can be seen that `regex` is not as good as `RE2` in processing regular expressions containing capturing groups. what is the reason?

What other regular expression `regex` performance is worse than `RE2`？

**Reply by BurntSushi** [reply]
> when I run "./run re2", I get the following error, how can I fix it

Install RE2? The build expects `pkg-config` to work on your system and expects `pkg-config` to tell Cargo where RE2 is. See: https://docs.rs/pkg-config/latest/pkg_config/

The benchmark build does integration with RE2 via a single line: https://github.com/rust-lang/regex/blob/159a63c85eb77ec321301bc4c4ebfb90343edc2b/bench/build.rs#L15

How to get RE2 wired up with `pkg-config` is beyond the scope of what I can help you with. Sorry. You'll have to seek help for that elsewhere.

> It can be seen that regex is not as good as RE2 in processing regular expressions containing capturing groups. what is the reason?

The reason is that your measurement process is flawed, and therefore any conclusion you draw from it is highly suspect. You can't compare benchmark results from one harness to the next. RE2 for example is not actually extracting any capture groups where as the Rust regex benchmark is. You need to use `find_iter` on the Rust regex side or adjust the RE2 benchmark to extract capture groups.

I implore you: please approach benchmarking like you would science. **Make what you're doing reproducible and ensure you're measuring the same thing.**

> What other regular expression regex performance is worse than RE2？

It's difficult to answer this. In general, I expect the regex crate to be at least as fast as RE2 in _most_ cases. It's hard to know every case where it isn't, but some surely exist. For example, I believe RE2 is a bit faster with regexes like `(\pL{10})|(\pL{9})|(\pL{8})|(\pL{7})|(\pL{6})|(\pL{5})` at present. Likely because it has better heuristics for when to cut over from the lazy DFA to the PikeVM, but I'm honestly not sure.

**Reply by BurntSushi** [reply]
OK, so I do have a WIP rewrite of the regex crate internals ongoing and as part of that work, I have a more sophisticated benchmark harness setup. So I took your `(?s)((.*)()()($))` regex and added it to the harness with one small tweak: I used `^(?s)((.*)()()($))` instead, since that's what the RE2 benchmark is measuring: it's running `FullMatch` which is an anchored search. The regex crate doesn't have "full match" APIs and instead forces one to use `^` instead. Here is the benchmark definition: https://github.com/BurntSushi/regex-automata/blob/babe9e0557b4904c745cf99b67556914bdeed44c/benchmarks/random.toml#L1-L104

I defined the full matrix of `{regex,hybrid,re2,pcre2-jit,pcre2-nojit} x {count,count-with-captures}`, where `regex` is the current regex crate and `hybrid` is the new lazy DFA. Assuming you've run `cargo install --path regex-cli` (and you might need to get rid of [these patches](https://github.com/BurntSushi/regex-automata/blob/babe9e0557b4904c745cf99b67556914bdeed44c/Cargo.toml#L26-L28) in order to build, I have those there because it's all WIP), then you can run the specific benchmarks with:

```
$ regex-cli bench measure -f random | tee 2022-08-26.csv
```

And then comparing them (just regex vs re2 vs hybrid):

```
$ regex-cli bench cmp -e '^(regex|re2)/api' -e hybrid 2022-08-26.csv
benchmark                                               re2/api     regex/api   regex/automata/hybrid
---------                                               -------     ---------   ---------------------
count-captures/random/dot-star-capture-ascii-lines-1mb  11.2 MB/s   10.9 MB/s   -
count-captures/random/dot-star-capture-ascii-lines-32   66.4 MB/s   68.6 MB/s   -
count-captures/random/dot-star-capture-ascii-lines-32k  11.2 MB/s   118.4 MB/s  -
count/random/dot-star-capture-ascii-lines-1mb           634.7 MB/s  317.4 MB/s  635.2 MB/s
count/random/dot-star-capture-ascii-lines-32            203.0 MB/s  185.1 MB/s  299.7 MB/s
count/random/dot-star-capture-ascii-lines-32k           633.4 MB/s  317.4 MB/s  634.4 MB/s
```

There are a few things worth observing here:

1. When it comes to actually resolving capturing groups, you can clearly see the big difference between doing it vs not doing it. This explains the big difference you observed in timings between RE2's benchmark and the regex crate's. In the RE2 benchmark, capturing groups aren't actually being matched, so it can be a lot faster. My new benchmark harness properly controls for all of that.
2. When you control for everything, RE2 and the regex crate are either roughly tied in speed or the regex crate is faster.
3. When it comes to just counting matches and _not_ getting capturing groups, the regex crate is actually twice as slow as RE2. That's because the current implementation doesn't notice that it doesn't need to run a reverse search to find the start position. Since the regex is anchored, it can skip that. The new rewrite does it, but the current crate doesn't. 

Popping up a level, the regex you chose here is honestly pretty weird. (The PCRE2 benchmarks show speeds that suggest it has an optimization that eliminates any haystack-proportional work at all, which is pretty cool.) If you care specifically about capturing groups, then there are other benchmarks that might be more interesting:

```
$ regex-cli bench cmp -e '^(regex|re2)/api' -e onepass -f count-captures -f grep-captures tmp/bench/all/2022-08-21.csv
benchmark                                                   re2/api      regex/api   regex/automata/onepass
---------                                                   -------      ---------   ----------------------
count-captures/internal/bibleref-long                       453.0 MB/s   452.2 MB/s  -
count-captures/internal/bibleref-short                      4.1 MB/s     20.7 MB/s   -
count-captures/internal/contiguous-letters                  -            670.2 KB/s  -
count-captures/internal/rust-functions                      3.8 GB/s     2.3 GB/s    -
count-captures/internal/two-or-three-long-words             630.6 MB/s   627.1 MB/s  -
grep-captures/grep/contiguous-letters                       2.8 MB/s     3.4 MB/s    -
grep-captures/grep/literal-star-variant1                    251.4 MB/s   329.6 MB/s  -
grep-captures/grep/literal-star-variant2                    253.0 MB/s   329.5 MB/s  -
grep-captures/grep/literal-star-variant3                    257.5 MB/s   328.9 MB/s  -
grep-captures/grep/overlapping-words-ascii                  39.2 MB/s    40.9 MB/s   -
grep-captures/grep/overlapping-words-russian                688.9 KB/s   215.7 KB/s  -
grep-captures/grep/overlapping-words-unicode                1122.7 KB/s  -           -
grep-captures/onepass/every-non-empty-line                  108.8 MB/s   140.9 MB/s  202.5 MB/s
grep-captures/onepass/every-non-empty-line-anchors          238.2 MB/s   166.6 MB/s  110.2 MB/s
grep-captures/onepass/first-three-words                     161.9 MB/s   33.0 MB/s   194.1 MB/s
grep-captures/onepass/first-three-words-boundaries-ascii    147.1 MB/s   31.7 MB/s   138.2 MB/s
grep-captures/onepass/first-three-words-boundaries-unicode  -            50.3 MB/s   114.3 MB/s
grep-captures/onepass/fn-first-param                        89.3 MB/s    106.0 MB/s  493.4 MB/s
grep-captures/onepass/fn-is-empty                           366.2 MB/s   859.7 MB/s  609.1 MB/s
```

I've included the `onepass` regex engine which is another new aspect of the rewrite, and shows that the regex crate will be improving quite a bit on most of the cases where RE2 is doing a lot better. But, it also shows where RE2 is better than the regex crate today. However, in many cases, perf is pretty competitive.

Once the new rewrite is in place, it should be easier to run `regex-cli`. But this is where I'm at for the moment.

**Reply by freekeeper** [reply]
As we know, `^` stands for start and `$` for end in regular expressions. So why use `^((.*)()()($))` instead of `^((.*)()()($))$`
 [https://github.com/BurntSushi/regex-automata/blob/babe9e0557b4904c745cf99b67556914bdeed44c/benchmarks/random.toml#L1-L104](https://github.com/BurntSushi/regex-automata/blob/babe9e0557b4904c745cf99b67556914bdeed44c/benchmarks/random.toml#L1-L104)

**Reply by BurntSushi** [reply]
Because it was a smaller change from the regex you were benchmarking with... And it's also semantically equivalent.

**Reply by freekeeper** [reply]
why fails when I make the following test:
```
macro_rules! begin_dot_star_capture_end {
    () => {
        r"^((.*)()()($))"
    };
}
bench_match!(is_match_begin_dot_star_capture_end, begin_dot_star_capture_end!(), get_text(TXT_1K, empty_suffix()));
```
The bench result is as follows:
```
test misc::is_match_begin_dot_star_capture_end ... FAILED
```
why is it successful when I make the following test:
```
macro_rules! dot_star_capture {
    () => {
        r"((.*)()()($))"
    };
}
bench_match!(is_match_dot_star_capture, dot_star_capture!(), get_text(TXT_1K, empty_suffix()));
```
The bench result is as follows:
```
test misc::is_match_dot_star_capture           ... bench:          24 ns/iter (+/- 19) = 42708 MB/s
```
What is the reason for the difference between the above two results?

**Reply by BurntSushi** [reply]
Ummm... I don't know? You need to give me a full repro. You haven't provided enough information for me to run what you're running. And I wouldn't necessarily expect them to be the same anyway. Your first regex is anchored. So it will match different things.

**Reply by freekeeper** [reply]
The added code is line 163 to 210 in this link [https://github.com/freekeeper/regex/blob/master/bench/src/misc.rs](https://github.com/freekeeper/regex/blob/master/bench/src/misc.rs)

**Reply by BurntSushi** [reply]
The answer is right here, in the docs for the macro: https://github.com/freekeeper/regex/blob/32bc38fa56bf8dc41bf23e8c46d50b69e3971805/bench/src/bench.rs#L124-L126

**Reply by freekeeper** [reply]
I did the following bench in regex，The complete code can be found at this [link](https://github.com/freekeeper/regex/blob/master/bench/src/misc.rs#L163-L210)
```
macro_rules! dot_star_capture {
    () => {
        r"((.*)()()($))"
    };
}
bench_match!(is_match_dot_star_capture, dot_star_capture!(), get_text(TXT_1K, empty_suffix()));
```
The bench result is as follows:
```
test misc::is_match_dot_star_capture           ... bench:          24 ns/iter (+/- 19) = 42708 MB/s
```
And I want to bench regex's interface to C's  `rure_is_match` [https://github.com/rust-lang/regex/blob/master/regex-capi/include/rure.h#L185](https://github.com/rust-lang/regex/blob/master/regex-capi/include/rure.h#L185) function with the same test text and regex, the core code is as follows:
```
void Rure_is_Match_RE2(benchmark::State& state, const char *regexp)
{
  std::ifstream in("re2/testing/text_re2_1KB.txt");
  std::stringstream buffer;
  buffer << in.rdbuf();
  std::string s = buffer.str().substr(0, state.range(0));
  rure_error *err = rure_error_new();
  rure *re1 = rure_compile((const uint8_t *)regexp, strlen(regexp), RURE_DEFAULT_FLAGS, NULL, err);
  for (auto _ : state) {
    bool matched = rure_is_match((rure *)re1, (const uint8_t *)s.c_str(), strlen(s.c_str()), 0);
    CHECK(matched);
  }
  state.SetBytesProcessed(state.iterations() * state.range(0));
}

void Rure_is_Match_RE2_DotStarCapture_text_re2_1KB(benchmark::State& state)  { Rure_is_Match_RE2(state, "(?s)((.*)()()($))"); }

BENCHMARK_RANGE(Rure_is_Match_RE2_DotStarCapture_text_re2_1KB, 2<<6, 2<<9);
```
The bench result is as follows:
```
Rure_is_Match_RE2_DotStarCapture_text_re2_1KB/128        5000000               385 ns/op         332.07 MB/s
Rure_is_Match_RE2_DotStarCapture_text_re2_1KB/256        5000000               382 ns/op         669.85 MB/s
Rure_is_Match_RE2_DotStarCapture_text_re2_1KB/512        5000000               378 ns/op        1354.12 MB/s
Rure_is_Match_RE2_DotStarCapture_text_re2_1KB/1K         5000000               385 ns/op        2659.34 MB/s
```
I use the same test text and the same regular expression, why is the performance so bad in `rure_is_match`

**Reply by BurntSushi** [reply]
Dunno. I mean, you are running `strlen` inside the benchmark, but I wouldn't expect that to fully explain the difference here.

The best I can say is that something about your measurement process is flawed. But you aren't really providing a way for me to easily reproduce your tests. For example, I'd expect, "clone this repo and run this script" and it should do exactly what you're trying. Otherwise, I'm not digging through RE2's benchmark harness to try and understand it for you. I've already explained what I could.
